lib/debug/server_dap.rb

```diff
@@ -745,8 +745,8 @@ def process_dap args
                   break false
                 end
               } and (message = "Error: Not defined global variable: #{expr.inspect}")
-            when /\A[A-Z]/
-              unless result = search_const(b, expr)
+            when /(\A[A-Z]\w*)/
```

Comment:
just curious, why don't we use `frame_eval` to evaluate the `expr` directly but instead parsing them?

Reply:
`frame_eval` sometimes changes the value(FYI: https://github.com/ruby/debug/pull/395).

Reply:
I think after https://github.com/ruby/debug/commit/756a2ed01d12a5fbdb81da829e48f159dd688ad6#diff-4208807b6db8555c61e42002b3e9c879003c5cfdb32d2b1a285a614b22fdf9d5 the changes made by `frame_eval` doesn't affect the original program anymore. So in this case it won't reproduce the issue again.

Reply:
> I think after 756a2ed#diff-4208807b6db8555c61e42002b3e9c879003c5cfdb32d2b1a285a614b22fdf9d5 the changes made by frame_eval doesn't affect the original program anymore. So in this case it won't reproduce the issue again.

Oh, I see. I didn't know that.
However, we can't use `frame_eval` for the following reasons:

- `frame_eval` uses current_frame, but we have to get a frame from request.
  - https://github.com/ruby/debug/blob/master/lib/debug/server_dap.rb#L715
- Even current_frame is the same as the target frame, we can't sometimes get the correct value such as instance of class. In the following case, showed value and value in backtrace information are not same.
  - <img width="519" alt="Screen Shot 2021-12-27 at 9 47 56 AM" src="https://user-images.githubusercontent.com/59436572/147424032-610be64f-dbfd-45b8-8dfd-ac5381e0e4a5.png">

Reply:
I don't get the example case. Are they different because they came from different frames? Or they're different but are evaluated from the same frame?
It looks like `search_const` may not benefit much from using `frame_eval`. But I think instance variables, global variables and normal expressions can be evaluated with `frame_eval`. And making `frame_eval` take arbitrary frame's binding shouldn't be an issue. 

I also see that the CDP server uses a similar but slightly different implementation for evaluation. So if we can utilize it in both server implementation, we can largely simplify these parts and make sure we're aligned on features (such as special locals).

Reply:
We shouldn't use `eval` because expression can have side-effect.